### ISSUE TITLE
Global var set before evaluating function declaration in non-strict block

### ISSUE THREAD

**Issue by zenparsing** [Microsoft/ChakraCore#5340]
For the following test case:

```js
function getF() { return f; }
{
  if (getF() === f) {
    print('Global var set before evaluating function declaration');
  }
  function f() { return "bar"; }
}

```

I see:

```
#### ch
Global var set before evaluating function declaration

#### sm


#### v8


#### jsc
Global var set before evaluating function declaration
```

According to [B3.3.2](https://tc39.github.io/ecma262/#sec-web-compat-globaldeclarationinstantiation), the global "var" binding should be initialized to undefined, and only set to the block-level value of f when the function declaration is evaluated.

**Comment by zenparsing** [comment]
Annex B3.3 is crazy stuff.

It says that if there's a function in a block scope in non-strict code, then when we begin executing the script we initialize a global var with that identifier to undefined. When we enter the block, we create a block-scoped, hoisted binding for the function. When we reach the function definition, we set the outer (global) var binding to the current value of the inner shadowing block scope binding.

I don't think there is a test262 test for this particular edge case.

**Comment by MSLaguana** [comment]
Ah; so the bug in this case is that we aren't distinguishing between the block-scoped binding of `f` and the global binding of `f`, we are just directly setting the global binding of `f` when we enter the block scope?

**Comment by pleath** [comment]
We set both the global and the block-scoped bindings at the same time. We almost certainly have the same issue with non-global code. It should be a pretty simple thing to fix. Does anyone actually implement the spec today?

**Comment by fatcerberus** [comment]
Based on the test results above it appears that both SpiderMonkey and V8 implement this properly.

**Comment by fatcerberus** [comment]
I feel it's worth noting that the reason Annex B exists at all is simply for the sake of legacy Web compatibility; not implementing it properly WILL break a lot of existing code (in fact it's my understanding that implementing Annex B is only required if you're writing a browser).  In this particular case, block-scoped functions are specified the way they are because many devs in the past would conditionally define functions like this:

```js
if (someCondition) {
    function doSomethingAwesome() {
        // do stuff
    }
}
```

If that function is hoisted onto the global object too early (as we appear to be doing now) then such code no longer does what the programmer intended.

**Comment by pleath** [comment]
You're right, the annex is crucial for backward compat. Our divergence from the spec is pretty subtle, though. We're doing the important thing correctly, which is to instantiate the function only if awesomenessIsAvailable is true and to let people call it later, from outside the block, by referring to the global. The fact that we're assigning the right value to the global on entry to the if block doesn't seem likely to break legacy code. We should certainly fix it, but @zenparsing 's assessment of severity 3 seems appropriate to me.
